docs/proposals/apiserver_watch.md

```diff
@@ -0,0 +1,132 @@
+## Abstract
+
+In the current system, all watch requests send to apiserver are in general
+redirected to etcd. This means that for every watch request to apiserver,
+apiserver opens a watch on etcd.
+
+The purpose of the proposal is to improve the overall performance of the system
+by solving the following problems:
+
+- having too many open watches on etcd
+- avoiding deserializing/converting the same objects multiple times in different
+watch results
+
+In the future, we would also like to add an indexing mechanism to the watch.
+Although Indexer is not part of this proposal, it is supposed to be compatible
+with it - in the future Indexer should be incorporated into the proposed new
+watch solution in apiserver without requiring any redesign.
+
+
+## High level design
+
+We are going to solve those problems by allowing many clients to watch the same
+storage in the apiserver, without being redirected to etcd.
+
+At the high level, apiserver will have a single watch open to etcd, watching all
+the objects (of a given type) without any filtering. The changes delivered from
+etcd will then be stored in a cache in apiserver. The introduced cache will have
+a limited capacity - whenever a new change comes from etcd and the cache is full,
+the oldest change will be removed to make place for the new change.
+
+When a client sends a watch request to apiserver, instead of redirecting it to
+etcd, it will just be registered in that cache. This will result in creating a
+go-routine that will be responsible for iterating over the cache, applying
+the matcher and sending filtered changes directly to the client.
+
+The following section describes the proposal in more details, analizes some
+corner cases and divides the whole design in more fine-grained steps.
+
+
+## Proposal details
+
+We would like the cache to be __per-resource-type__ and __optional__. Thanks to
+it we will be able to:
+  - have different cache sizes for different resources (e.g. bigger cache
+    [= longer history] for pods, which can significantly affect performance)
+  - avoid any overhead for objects that are watched very rarely (e.g. events
```

Comment:
We'll have all the state in memory, so we should also be able to return resources (e.g. pods) without new calls to etcd.